baremetal/metal3machine_manager.go

```diff
@@ -1259,122 +1275,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
 // ClientGetter prototype.
 type ClientGetter func(ctx context.Context, c client.Client, cluster *clusterv1.Cluster) (clientcorev1.CoreV1Interface, error)
 
-// SetNodeProviderID sets the metal3 provider ID on the kubernetes node.
-func (m *MachineManager) SetNodeProviderID(ctx context.Context, providerIDOnM3M *string, clientFactory ClientGetter) error {
-	corev1Remote, err := clientFactory(ctx, m.client, m.Cluster)
+func (m *MachineManager) setNodeProviderID(ctx context.Context, client clientcorev1.CoreV1Interface, node corev1.Node, providerID string) error {
+	oldData, err := json.Marshal(node)
 	if err != nil {
-		return errors.Wrap(err, "Error creating a remote client")
+		return fmt.Errorf("failed to json.Marshal node: %w", err)
+	}
+
+	node.Spec.ProviderID = providerID
+
+	nodeVar := node.DeepCopy()
+	newData, err := json.Marshal(*nodeVar)
+	if err != nil {
+		return fmt.Errorf("failed to json.Marshal node: %w", err)
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, corev1.Node{})
+	if err != nil {
+		return fmt.Errorf("failed to create patch for node %q: %w", node.GetName(), err)
+	}
+	_, err = client.Nodes().Patch(ctx, nodeVar.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to update the target node with providerID: %w", err)
+	}
+
+	return nil
+}
+
+func (m *MachineManager) getMetal3MachineHostnames() []string {
+	hostnames := []string{}
+
+	for _, address := range m.Metal3Machine.Status.Addresses {
+		if address.Type == clusterv1.MachineHostName {
+			hostnames = append(hostnames, address.Address)
+		}
 	}
+
+	return hostnames
+}
+
+// SetProviderID sets the metal3 provider ID on the Metal3Machine.
+func (m *MachineManager) SetProviderID(providerID string) {
+	m.Log.Info("ProviderID set on the Metal3Machine", "providerID", providerID)
+	m.Metal3Machine.Spec.ProviderID = &providerID
+}
+
+// SetDefaultProviderID sets the ProviderID on this Metal3Machine using the new format.
+func (m *MachineManager) SetDefaultProviderID() error {
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+	bmhName, err := m.getBmhNameFromM3Machine()
+
+	if err != nil {
+		return err
+	}
+
+	providerID := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	m.SetProviderID(providerID)
+	return nil
+}
+
+// getPossibleProviderIDs returns the ProviderID for this Metal3Machine in the legacy and the new format.
+func (m *MachineManager) getPossibleProviderIDs(ctx context.Context) (providerIDLegacy string, providerIDNew string, err error) {
 	namespace := m.Metal3Machine.GetNamespace()
 	m3mName := m.Metal3Machine.GetName()
 	bmhName, err := m.getBmhNameFromM3Machine()
 	if err != nil {
 		errMessage := "unable to retrieve BMH name from Metal3Machine"
 		m.Log.Info(errMessage)
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+		err = errors.Wrap(err, errMessage)
```

Comment:
why are we not returning `WithTransientError` anymore here ?

Reply:
Good question!

My thinking was that this code is now in a private helper function and we should reserve returning `WithTransientError` to public API functions that are called directly by the controller.  I'm happy to change it if you'd prefer.

Reply:
I agree, it looks like it's handled on the higher level.

Reply:
nit: One day, we should get rid of errors.Wrap and other bits from the deprecated 3rd party errors package.